docs/proposals/node-allocatable.md

```diff
@@ -0,0 +1,177 @@
+2. For external schedulers (such as mesos, hadoop, etc.) integration, they might want to partition
+   compute resources on a given node, limiting how much Kubelet can use. We should provide a
+   mechanism by which they can query kubelet, and reserve some resources for their own purpose.
+
+### Scope of proposal
+
+This proposal deals with resource reporting through the [`Allocatable` field](#allocatable) for more
+reliable scheduling, and minimizing resource over commitment. This proposal *does not* cover
+resource usage enforcement (e.g. limiting system component usage), and pod eviction (e.g. when
+reservation grows).
+
+## Design
+
+### Definitions
+
+![image](node-allocatable.png)
+
+1. **Machine Capacity** - Already provided as
+   [`NodeStatus.Capacity`](https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/v1/definitions.html#_v1_nodestatus),
+   this is total capacity read from the machine instance, and assumed to be constant.
+2. **Reserved** (proposed) - Compute resources reserved for non-Kubernetes and non-system
+   workloads. For example, processes scheduled to run alongside Kubernetes on Mesos.
+3. **Kubelet Allocatable** - Compute resources available for scheduling (including scheduled &
+   unscheduled resources). This value is currently implicit, and the focus of this proposal. See
+   [below](#api-changes) for more details.
+4. **System Reserved** (proposed) - Compute resources reserved for system components such as the docker daemon, kubelet, kube proxy, etc.
+
+### API changes
+
+#### Allocatable
+
+Add `Allocatable` (4) to
+[`NodeStatus`](https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/v1/definitions.html#_v1_nodestatus):
+
+```
+type NodeStatus struct {
+  ...
+  // Allocatable represents schedulable resources of a node.
+  Allocatable ResourceList `json:"allocatable,omitempty"`
+  ...
+}
+```
+
+Allocatable will be computed by the Kubelet and reported periodically to the API server. It is
+defined to be:
+
+```
+   [Allocatable] = [Machine Capacity] - ( [Reserved] + [System Reserved] + [Kernel usage (smoothed)] )
+```
+
+The scheduler will use `Allocatable` in place of `Capacity` when scheduling pods, and the Kubelet
+will use it when performing admission checks.
+
+- Since kernel usage can fluctuate and is out of kubernetes control, `Allocatable` may vary over
+  time.
+
+#### System Reserved
+
+`SystemReserved` is the parameter specifying resources reserved for system components (4). It is
+provided as a command-line flag to the Kubelet at startup, and therefore cannot be changed during
+normal Kubelet operation (this will change in the [future](#future-work)).
+
+The flag will be specified as a serialized `ResourceList`, with resources defined by the API
+`ResourecName` and values specified in `resource.Quantity` format, e.g.:
+
+```
+--system-reserved=cpu:500m,memory:5Mi
```

Comment:
The --runtime-config flag in kube-apiserver is an example.

On Tue, Nov 17, 2015 at 1:07 PM, Tim St. Clair <notifications@github.com>
wrote:

> In docs/proposals/node-allocatable.md
> <https://github.com/kubernetes/kubernetes/pull/17201#discussion_r45120669>
> :
>
> > +will use it when performing admission checks.
> > +
> > +- Since kernel usage can fluctuate and is out of kubernetes control, `Allocatable` may vary over
> > +  time.
> > +
> > +#### System Reserved
> > +
> > +`SystemReserved` is the parameter specifying resources reserved for system components (4). It is
> > +provided as a command-line flag to the Kubelet at startup, and therefore cannot be changed during
> > +normal Kubelet operation (this will change in the [future](#future-work)).
> > +
> > +The flag will be specified as a serialized `ResourceList`, with resources defined by the API
> > +`ResourecName` and values specified in `resource.Quantity` format, e.g.:
> > +
> > +```
> > +--system-reserved=cpu:500m,memory:5Mi
>
> Fixed. Can you point me to an example of other map flags?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17201/files#r45120669>.
>